internal/blobstore/multipart.go

```diff
@@ -313,30 +314,100 @@ func (s *Store) setUploadHash(uploadId string, udoc *uploadDoc) (string, error)
 func (s *Store) copyBlob(w io.Writer, name string) error {
 	rc, _, err := s.Open(name, nil)
 	if err != nil {
-		return errgo.Notef(err, "cannot open blob %q", name)
+		return errgo.NoteMask(err, fmt.Sprintf("cannot open blob %q", name), errgo.Is(ErrNotFound))
 	}
 	defer rc.Close()
 	if _, err := io.Copy(w, rc); err != nil {
-		return errgo.Notef(err, "cannot read blob %q", name)
+		if errgo.Cause(err) == mgo.ErrNotFound {
+			return errgo.WithCausef(err, ErrNotFound, "error reading blob %q", name)
+		}
+		return errgo.Notef(err, "error reading blob %q", name)
 	}
 	return nil
 }
 
-// DeleteExpiredUploads deletes any multipart entries
+// RemoveExpiredUploads deletes any multipart entries
 // that have passed their expiry date.
-func (s *Store) DeleteExpiredUploads() error {
-	return errgo.New("not implemented yet")
+func (s *Store) RemoveExpiredUploads() error {
+	it := s.uploadc.Find(bson.D{
+		{"expires", bson.D{{"$lt", time.Now()}}},
+	}).Iter()
+	defer it.Close()
+	var udoc uploadDoc
+	for it.Next(&udoc) {
+		err := s.deleteUpload(&udoc)
+		if err != nil {
+			return errgo.Mask(err)
+		}
+	}
+	if err := it.Err(); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
 }
 
-// DeleteUpload deletes all the parts associated with the
+// RemoveUpload deletes all the parts associated with the
 // given upload id. It is a no-op if called twice on the
 // same upload id.
-func (s *Store) DeleteUpload(uploadId string) error {
-	// TODO
-	// read multipart metadata
-	// delete all parts referenced in that
-	// delete multipart metadata
-	return errgo.New("not implemented yet")
+func (s *Store) RemoveUpload(uploadId string) error {
+	udoc, err := s.getUpload(uploadId)
+	if err != nil {
+		if errgo.Cause(err) == ErrNotFound {
+			return nil
+		}
+		return errgo.Mask(err)
+	}
+	return s.deleteUpload(udoc)
+}
+
+func (s *Store) deleteUpload(udoc *uploadDoc) error {
+	removedDoc := false
+	removeParts := udoc.Hash == ""
+	if removeParts {
+		// It's possible that FinishUpload has been called
+		// at the same time as RemoveUpload, so only
+		// remove the upload document if that hasn't
+		// happened.
+		err := s.uploadc.Remove(bson.D{
+			{"_id", udoc.Id},
+			{"hash", bson.D{{"$exists", false}}},
+		})
+		switch err {
+		case nil:
+			removedDoc = true
+		case mgo.ErrNotFound:
+			// Someone called FinishUpload concurrently.
+			removeParts = false
+		default:
+			return errgo.Mask(err)
+		}
+	}
+	if !removedDoc {
+		if err := s.uploadc.RemoveId(udoc.Id); err != nil && err == mgo.ErrNotFound {
```

Comment:
err == mgo.ErrNotFound implies that it isn't nil doesn't it? Although I have a suspicion that's not the operator you wanted.